DetectorDescription/Core/interface/Store.h

```diff
@@ -9,136 +9,129 @@
-         // ELSE     
-         result.first->second = new Rep_type(n,(I)nullptr);
+      auto result = reg_.emplace( n, tmp );
+      if( result.second ) {
+	if( readOnly_ )
+	  throw cms::Exception( "DetectorDescriptionStore" )
+	    << " Store has been locked. Illegal attempt to add "
+	    << n << " to a global store."; 
+	// ELSE     
+	result.first->second = new Rep_type( n, nullptr );
       }
       return result.first->second;    
-   }
-      
-   template<class N, class I, class K>
-   typename Store<N,I,K>::prep_type 
-   Store<N,I,K>::create(const name_type & n, 
-                        pimpl_type p)
-   {			
-      if (readOnly_) throw cms::Exception("DetectorDescriptionStore")<<" Store has been locked.  Illegal attempt to add " << n << " to a global store."; 
-      // ELSE     
+    }
+  
+  template<class N, class I, class K>
+    typename Store<N,I,K>::prep_type
+    Store<N,I,K>::create( const name_type & n,
+			  pimpl_type p )
+    {			
+      if( readOnly_ )
+	throw cms::Exception( "DetectorDescriptionStore" )
+	  << " Store has been locked. Illegal attempt to add "
+	  << n << " to a global store."; 
+      // ELSE
       prep_type tmp = nullptr;
-      auto result = reg_.insert(std::make_pair(n,tmp));
-      if (!result.second) {
-         delete result.first->second->second;
-         result.first->second->second = p;
+      auto result = reg_.emplace( n, tmp );
+      if( !result.second ) {
+	result.first->second->second.swap( p );
       }
       else {
-         result.first->second = new Rep_type(n,p);
+	result.first->second = new Rep_type( n, std::move( p ));
       }
       return result.first->second;
-   }
-   
-   template<class N, class I, class K>
-   Store<N,I,K>::~Store()
-   {
-      for( auto it : reg_ ) {
-         delete it.second->second;
-	 it.second->second = nullptr;
-         delete it.second;
-	 it.second = nullptr;
-      }
-   } 
-   
-   template<class N, class I, class K>
-   bool Store<N,I,K>::isDefined(const name_type & n ) const
-   {
+    }
+  
+  template<class N, class I, class K>
+    bool Store<N,I,K>::isDefined(const name_type & n ) const
+    {
       if (readOnly_) edm::LogWarning("DetectorDescriptionStore") << " Store is locked and most likely empty.  isDefined will be false.";
       auto it = reg_.find(n);
       bool result(false);
       if (it != reg_.end()) {
-         if (it.second->second) {
-            result=true;
-         }  
+	if (it.second->second) {
+	  result=true;
+	}  
       }
       return result;
-   }
-   
-   template<class N, class I, class K>
-   void Store<N, I, K>::swap ( Store<N, I, K>& storeToSwap ) {
-      reg_.swap(storeToSwap.reg_);
-      storeToSwap.readOnly_ = readOnly_;
-   }
-   
+    }
+  
+  template<class N, class I, class K>
+    void Store<N, I, K>::swap ( Store<N, I, K>& storeToSwap ) {
+    reg_.swap(storeToSwap.reg_);
```

Comment:
Can one swap values on a read-only object? Is that appropriate?

Reply:
No. Unfortunately, the readOnly flag is always set to false and does not prevent internal modifications of the Store.